crypto/encode_decode/decoder_lib.c

```diff
@@ -511,11 +512,19 @@ static int decoder_process(const OSSL_PARAM params[], void *arg)
 
         /* Recurse */
         new_data.current_decoder_inst_index = i;
+        ERR_set_mark();
         ok = new_decoder->decode(new_decoder_inst->decoderctx,
                                  (OSSL_CORE_BIO *)bio,
                                  decoder_process, &new_data,
                                  ossl_pw_passphrase_callback_dec,
                                  &new_data.ctx->pwdata);
+        /* prune low-level ASN.1 parse errors from error queue */
+        err = ERR_peek_last_error();
+        if (ERR_GET_LIB(err) == ERR_LIB_ASN1
+                && ERR_GET_REASON(err) == ASN1_R_HEADER_TOO_LONG)
+            ERR_pop_to_mark();
+        else
+            ERR_clear_last_mark();
```

Comment:
Aha, yeah, so apparently, what I said in https://github.com/openssl/openssl/pull/12893#issuecomment-693447764 didn't quite catch.  I repeat, libcrypto _can't make assumptions on what errors are recorded_.  This kind of dropping is better done in the implementation (i.e. `providers/implementations/encode_decode/*`.

This gets a firm -1 from me

Reply:
I am very aware that reacting on error queue entries often involves guesswork and can break.
Yet I do not see much other chance here to get rid of
```
C00777D9AA7F0000:error::asn1 encoding routines:(unknown function):header too long:crypto/asn1/asn1_lib.c:103:
```
which comes form another part of the crypto lib, not a volatile external implementation.

BTW, at least in the code mentioned above as part of  commit c2150f73571 you did the same :wink:
```
        if (ERR_GET_LIB(err) == ERR_LIB_ASN1                            \
            && ERR_GET_REASON(err) == ERR_R_NESTED_ASN1_ERROR)          \
            ERR_pop_to_mark();                                          \
```

Reply:
> BTW, at least in the code mentioned above as part of commit c2150f7 you did the same :wink:

And that's code that make internal libcrypto calls from within libcrypto.
Your code is doing this deliberately around a call the crosses the libcrypto \<-\> provider boundary.
That's a huge difference, and I'm apparently not able to get the the implications across to you.

Reply:
I see the difference, and I acknowledge that arbitrary crypto-external provider code could have been called,
but as mentioned in this case it is not actually provider code but ASN.1 crypto module code,
as can be seen also from `ERR_GET_LIB(err) == ERR_LIB_ASN1` (and code outside the crypto lib should dare to use `ASN1err()`).

Do you have a more elegant suggestion for getting rid of the following offending error queue entry?
```
C00777D9AA7F0000:error::asn1 encoding routines:(unknown function):header too long:crypto/asn1/asn1_lib.c:103:
```

Reply:
> Do you have a more elegant suggestion for getting rid of the following offending error queue entry?

As a matter of fact, I do, and that's to prune them away around spots like this:

https://github.com/openssl/openssl/blob/ebcae87f6b62ed9b79a1255dbb3c69c635cea4d8/providers/implementations/encode_decode/encode_key2any.c#L212

... or perhaps deeper inside those functions.  Essentially, close enough to the actual i2d call (look for the variable k2d)

Reply:
That's what meant when I wrote "This kind of dropping is better done in the implementation (i.e. `providers/implementations/encode_decode/*`."

Reply:
> > Do you have a more elegant suggestion for getting rid of the following offending error queue entry?
> 
> As a matter of fact, I do, and that's to prune them away around spots like this:
> 
> https://github.com/openssl/openssl/blob/ebcae87f6b62ed9b79a1255dbb3c69c635cea4d8/providers/implementations/encode_decode/encode_key2any.c#L212
> 
> ... or perhaps deeper inside those functions. Essentially, close enough to the actual i2d call (look for the variable k2d)

This PR is about decoding, not decoding, so `d2i` rather than `i2d`.
Following your suggestion, I've move the error queue pruning from `decoder_process()` to 
`providers/implementations/encode_decode/decode_der2key.c` and
`b/providers/implementations/storemgmt/file_store_der2obj.c`
although I am not happy about it.

Reply:
> That's what meant when I wrote "This kind of dropping is better done in the implementation (i.e. `providers/implementations/encode_decode/*`."

The approach you suggested has two drawbacks:
* one needs to detect and manipulate all spots where a decoding attempt initiated by `decoder_process()` is done
* the changed functions can be called also in other situations where it may not be adequate to prune parse errors.

Reply:
> - the changed functions can be called also in other situations where it may not be adequate to prune parse errors.

I'm lost. In what case are the decoding functions in the provider not ultimately called from `decoder_process`?

Reply:
Ah, yes, those functions are static and access to them appears pretty indirect,
such that it would be hard to (ab-)use them from elsewhere. 
So I retract my second point of criticism.

Reply:
That should also resolve the first point.  There are certain functions, basically the main drivers in `providers/implementations/encode_decode/decode*.c`, where you *know* that the call comes from `decoder_process`, and you can conclude that there's an attempt to guess, and can therefore safely prune errors that are simply a guess that failed.  So basically, it would be the same pruning you already do, but around these three calls:

https://github.com/openssl/openssl/blob/28833f1465a2dd197f8df80a69095d1913e6e85e/providers/implementations/encode_decode/decode_der2key.c#L184
https://github.com/openssl/openssl/blob/28833f1465a2dd197f8df80a69095d1913e6e85e/providers/implementations/encode_decode/decode_der2key.c#L188
https://github.com/openssl/openssl/blob/28833f1465a2dd197f8df80a69095d1913e6e85e/providers/implementations/encode_decode/decode_der2key.c#L193

Reply:
>  That should also resolve the first point. There are certain functions, basically the main drivers in providers/implementations/encode_decode/decode*.c, where you know that the call comes from decoder_process, and you can conclude that there's an attempt to guess, and can therefore safely prune errors that are simply a guess that failed. So basically, it would be the same pruning you already do, but around these three calls:

The three spots/calls you mentioned are already part of where I added the pruning, but they are not sufficient.
I found that spurious parse errors also also due to the calls before (at the beginning of `der2key_decode()`) as well as the call `asn1_d2i_read_bio()` in `der2obj_decode()` had to be pruned.
Hope that I found all relevant spots, or are there maybe even more to cover?

This actually strengthens my first point of criticism that it's hard/error-prone to find all spots where the pruning is needed;
moreover it's not elegant to spread the very same pruning over several places (rather than having it simply at the central spot in `decoder_process()` of the crypto lib).

Reply:
It took me again several hours to find out how to prune further ASN.1 errors for the 'minimal' configuration
`--strict-warnings no-shared no-dso no-pic no-aria no-async no-autoload-config no-blake2 no-bf no-camellia no-cast no-chacha no-cmac no-cms no-cmp no-comp no-ct no-des no-dgram no-dh no-dsa no-dtls no-ec2m no-engine no-filenames no-gost no-idea no-ktls no-mdc2 no-md4 no-multiblock no-nextprotoneg no-ocsp no-ocb no-poly1305 no-psk no-rc2 no-rc4 no-rmd160 no-seed no-siphash no-siv no-sm2 no-sm3 no-sm4 no-srp no-srtp no-ssl3 no-ssl3-method no-ts no-ui-console no-whirlpool no-asm`